graylog2-web-interface/src/views/components/visualizations/ChartData.js

```diff
@@ -10,64 +10,81 @@ export type ChartDefinition = {
   name: string,
   x?: Array<string>,
   y?: Array<any>,
+  z?: Array<Array<any>>,
 };
 
-export type ChartData = [any, Array<Key>, Array<any>]
+export type ChartData = [any, Array<Key>, Array<any>, Array<Array<any>>]
 export type ExtractedSeries = Array<ChartData>;
 
 export type KeyJoiner = (Array<any>) => string;
 
-export type Generator = (string, string, Array<string>, Array<any>, number, number) => ChartDefinition;
+export type Generator = (string, string, Array<string>, Array<any>, Array<Array<any>>, number, number) => ChartDefinition;
 
 const _defaultKeyJoiner = keys => keys.join('-');
 
-export const extractSeries = (keyJoiner: KeyJoiner = _defaultKeyJoiner) => {
-  return (results: Rows): ExtractedSeries => {
-    // $FlowFixMe: Somehow flow is unable to infer that the result consists only of Leafs.
-    const leafs: Array<Leaf> = results.filter(row => (row.source === 'leaf'));
+const _defaultChartGenerator = (type, name, labels, values): ChartDefinition => ({ type, name, x: labels, y: values });
 
-    const x = leafs.map(({ key }) => key);
+const _flattenLeafs = (leafs: Array<Leaf>, showSummarySeries: boolean = true) => {
+  const filterCondition = (value) => {
+    if (!value.source.endsWith('leaf')) {
+      return false;
+    }
+    if (!showSummarySeries) {
+      return value.source !== 'row-leaf';
+    }
+    return true;
+  };
+  return flatten(leafs.map(l => l.values.filter(v => filterCondition(v)).map(v => [l.key, v])));
+};
 
-    const y = flatten(leafs.map(l => l.values.filter(v => v.source.endsWith('leaf')).map(v => [l.key, v])));
+export const formatSeries = ({ valuesBySeries, xLabels }: {valuesBySeries: Object, xLabels: Array<any>}): ExtractedSeries => {
+  return Object.keys(valuesBySeries).map(value => [
+    value,
+    xLabels,
+    valuesBySeries[value],
+    [],
```

Comment:
I had to add a default value here, I tried to make the z value optional, in the generateChart parameter type definition, but got a flow error, that this (passing null, undefined) is not possible for a value inside an array.